src/Core.hs

```diff
@@ -63,14 +63,22 @@ data TypePattern
   | AnyType Ident Var
   deriving (Data, Eq, Show)
 
-data ConstructorPattern
-  = ConstructorPattern !Constructor [(Ident, Var)]
-  | AnyConstructor Ident Var
+data ConstructorPatternF pat
+  = CtorPattern !Constructor [pat]
+  | PatternVar Ident Var
+  deriving (Data, Eq, Foldable, Functor, Show, Traversable)
+makePrisms ''ConstructorPatternF
+
+newtype ConstructorPattern =
+  ConstructorPattern { _constructorPattern :: ConstructorPatternF ConstructorPattern }
   deriving (Data, Eq, Show)
-makePrisms ''ConstructorPattern
+makeLenses ''ConstructorPattern
```

Comment:
Since `ConstructorPattern` is a newtype, `makePrisms` and `makeLenses` both create an `Iso`, the only difference is the name; `makePrism` calls it `_ConstructorPattern`, while `makeLenses` names it `constructorPattern`. I like the former, because this allows me to use the more conventional name `unConstructorPattern` for the field name.

Reply:
Sounds good! Though I don't get the difference here - why does `unConstructorPattern` clash with `constructorPattern`?

Reply:
It's not a clash between a field named `unFoo` and a lens named `foo`, it's that if the field name is `unFoo` instead of `_foo`, then it doesn't begin with an underscore and so no lens will be generated.